chromadb/test/property/test_cross_version_persist.py

```diff
@@ -0,0 +1,194 @@
+import sys
+import os
+import shutil
+import subprocess
+import tempfile
+from typing import Generator, Tuple
+from hypothesis import given
+import pytest
+import json
+from urllib import request
+from chromadb.api import API
+from chromadb.test.configurations import (
+    persist_old_version_configurations,
+)
+import chromadb.test.property.strategies as strategies
+import chromadb.test.property.invariants as invariants
+from importlib.util import spec_from_file_location, module_from_spec
+from packaging import version as packaging_version
+import re
+import multiprocessing
+from chromadb import Client
+from chromadb.config import Settings
+
+MINIMUM_VERSION = "0.3.20"
+COLLECTION_NAME_LOWERCASE_VERSION = "0.3.21"
+version_re = re.compile(r"^[0-9]+\.[0-9]+\.[0-9]+$")
+
+
+def versions():
+    """Returns the pinned minimum version and the latest version of chromadb."""
+    url = "https://pypi.org/pypi/chromadb/json"
+    data = json.load(request.urlopen(request.Request(url)))
+    versions = list(data["releases"].keys())
+    # Older versions on pypi contain "devXYZ" suffixes
+    versions = [v for v in versions if version_re.match(v)]
+    versions.sort(key=packaging_version.Version)
+    return [MINIMUM_VERSION, versions[-1]]
+
+
+test_old_versions = versions()
+base_install_dir = tempfile.gettempdir() + "/persistence_test_chromadb_versions"
+
+
+def get_path_to_version_install(version):
+    return base_install_dir + "/" + version
+
+
+def get_path_to_version_library(version):
+    return get_path_to_version_install(version) + "/chromadb/__init__.py"
+
+
+def install_version(version):
+    # Check if already installed
+    version_library = get_path_to_version_library(version)
+    if os.path.exists(version_library):
+        return
+    path = get_path_to_version_install(version)
+    install(f"chromadb=={version}", path)
+
+
+def install(pkg, path):
+    # -q -q to suppress pip output to ERROR level
+    # https://pip.pypa.io/en/stable/cli/pip/#quiet
+    print(f"Installing chromadb version {pkg} to {path}")
+    return subprocess.check_call(
+        [
+            sys.executable,
+            "-m",
+            "pip",
+            "-q",
+            "-q",
+            "install",
+            pkg,
+            "--target={}".format(path),
+        ]
+    )
+
+
+def switch_to_version(version):
+    module_name = "chromadb"
+    # Remove old version from sys.modules, except test modules
```

Comment:
@levand This installs all the dependencies in the folder as well. Since we insert this folder to the path in a clean python environment, these dependencies are used.

<img width="784" alt="deps_isolated" src="https://user-images.githubusercontent.com/5598697/234060130-f6b6c9df-9ee4-4482-b90e-0b82444cd508.png">